Move listening to ListenService

fwiesel opened the pull request:
Since I was running also into #6, I thought I give it a spin.

The diff is much bigger than needed, and I can cut it manually down into smaller commits, and remove 
changes, if undesired.

I just used android studio in the latest version, and it was raising various "warnings" on what to change.
Since that is my first step into Android code, I was unclear what was really needed to make it work,
I just went with it.

Is there a way / tool to automatically enforce your desired coding style?

enguerrand commented:
First of all: Thanks a lot for stepping in to work on this issue. I did already experiment with this offline but ran into various issues, but maybe you already solved these or we can solve them together. 
At this point let's not worry about structure and code style too much but first get it to work properly and then refactor for styling and structure afterwards.
To that end I just created the branch feature/listen-service. 
Would you mind rebasing your PR onto this branch? This will make it easier to collaborate on this until it is in a state that I can merge into master.
Thanks again, I really appreciate the support!

enguerrand commented:
I just tested this (also android 12 so far) and it look a lot better than what I had so far, so this is great. 

One problem I did find is this: When I close the app by hitting the menu button the notification also disappears from the tray. If I then do an upwards swipe (to open the choice between different apps) and then discard child-monitor by swiping it up, the listen service also ends.

The behavior I was trying to build (unsuccessfully) in my own attempts was this: 

- Keep the notification tray entry at all times as long as the listening is not actively ended by the user
- open the running listen activity when clicking the notification tray entry
- only end the listen service when actively closing the listen activity in the child monitor ui.

I am not saying that it has to be like this but I'd be interested in your input on the above. Would you agree that this would be a desirable behavior and, if so, do you have an idea of how to achieve this?
My current understanding after reading the docs a bit is that if I want to bind a service to an activity then I cannot keep it running when the last activity unbinds. On the other hand, when the activity is "destroyed" it seems more clean to unbind. 
Does this make sense?

fwiesel commented on gradle.properties:
```diff
@@ -22,4 +22,5 @@
-android.nonTransitiveRClass=false
\ No newline at end of file
+android.nonTransitiveRClass=false
+org.gradle.unsafe.configuration-cache=true
```
As I read it, the setting apparently is outdated, and is now called `org.gradle.configuration-cache` (I've fixed the commit in that direction).
In the past you had to run a check to make sure that it is safe to use that option, now gradle itself will complain and bail out.

But I agree, it is not really a bit project, so it isn't needed. If you still want to leave it rather out, tell me, and I'll drop it.

fwiesel commented on app/src/main/java/de/rochefort/childmonitor/ListenService.java:
```diff
@@ -0,0 +1,250 @@
+import java.util.ArrayList;
+
+public class ListenService extends Service {
+    static final int MSG_REGISTER_CLIENT = 1;
```
Yes, sorry. That's from an earlier attempt, where I followed an example with multiple processes.
I've removed them.

fwiesel commented on app/src/main/java/de/rochefort/childmonitor/ListenActivity.java:
```diff
@@ -16,169 +16,125 @@
-        _mNotifyMgr =
-                NotificationManagerCompat.from(this);
+        final Bundle bundle = getIntent().getExtras();
+        if (bundle != null) {
```
True, I check now the error, and write a status message (and do not start the service).

fwiesel replied:
I suspect, we have to store the state in `onSaveInstanceState`  and restore it from the argument in that case.

enguerrand replied:
I just committed something that seems to work, let me know what you think

fwiesel replied:
Looking good. Probably better than the alternative I had: Copying the variables to the Intent passed to the pending intend. But your way, the state is stored only in one place.

fwiesel commented on app/src/main/java/de/rochefort/childmonitor/ListenActivity.java:
```diff
@@ -16,169 +16,125 @@
+    private boolean shouldUnbind;
+
+    // To invoke the bound service, first make sure that this value
+    // is not null.
```
Correct, I've fixed it. Instead of accessing the variable, I'll use 
`context.stopService(intent)`, which incidentally also fixes that the service continues running in the background when you navigate back.

fwiesel replied:
Main reason is, that I don't know if it won't get garbage collected, if I don't keep a reference around.

enguerrand replied:
Hm, it would surprise me if a service depended on an activity to prevent it from being garbage collected.

fwiesel commented on app/src/main/java/de/rochefort/childmonitor/ListenActivity.java:
```diff
@@ -16,169 +16,125 @@
+            // interact with the service.  Because we have bound to a explicit
+            // service that we know is running in our own process, we can
+            // cast its IBinder to a concrete class and directly access it.
+            boundService = ((ListenService.ListenBinder) service).getService();
```
I followed that pattern here and in other places.

fwiesel commented on app/src/main/java/de/rochefort/childmonitor/ListenActivity.java:
```diff
@@ -16,169 +16,125 @@
     protected void onStop() {
-        _listenThread.interrupt();
-        _listenThread = null;
+        boundService.stopForeground(true);
```
Done.

fwiesel commented on app/src/main/java/de/rochefort/childmonitor/ListenService.java:
```diff
@@ -0,0 +1,250 @@
+                client.send(Message.obtain(null, what));
+            }
+            catch (RemoteException e) {
+                // The client is dead. Remove it from the list; we are going through the list from back to front so this is safe to do inside the loop.
```
Yeah, I removed that. It is also from an example where the service was running independently from the app.
Sorry for the noise.

fwiesel commented on app/src/main/java/de/rochefort/childmonitor/ListenActivity.java:
```diff
@@ -16,169 +16,125 @@
+            final VolumeView volumeView = findViewById(R.id.volume);
+
+            volumeView.setVolumeHistory(boundService.getVolumeHistory());
+            boundService.setUpdateCallback(volumeView::postInvalidate);
```
I've changed it in a way, that all cross-thread used variables in `VolumeHistory` are now protected by a `synchronized` function.
Probably not the most performant solution, but I think if we are copying like 1k doubles, so 8kB which shouldn't be too bad.
The alternative would be to get a ringbuffer implementation with a "view" of a part of the ring-buffer, and block overwriting the "view".

fwiesel commented on app/src/main/java/de/rochefort/childmonitor/ListenService.java:
```diff
@@ -0,0 +1,250 @@
+
+    @Override
+    public void onDestroy() {
+        listenThread.interrupt();
```
It shows, I never close the thing :).
I check now for the listener being set to `null` (with copying it to a local variable), and fixed the stopping by sending an intent.

fwiesel commented on app/src/main/java/de/rochefort/childmonitor/VolumeHistory.java:
```diff
@@ -0,0 +1,40 @@
+    private double _volumeNorm = 1.0 / _maxVolume;
+    private final LinkedList<Double> _history = new LinkedList<>();
+
+    public double getLastNormalizedVolume() {
```
I've solved it probably a bit different as you intended, or at least differently than I understood you.
I copy the `max(size,width)` number of historic elements, plus the maximum/norm atomically out, and then work with that instead.
I hope, that works for you as well.

fwiesel commented:
> One problem I did find is this: When I close the app by hitting the menu button the notification also disappears from the tray. 
> Would you agree that this would be a desirable behavior and, if so, do you have an idea of how to achieve this?

Personally, I prefer it that way. Closing the monitor is simply closing the app. 

As I read it, the notification reflects the state of the service. And the only way to make the service persist over closing the application is making it an independent process from the app.
The implication here is, that every data-transfer between the service and the app becomes an IPC.

You stumbled over some of the left-overs from that attempt in your review.
The example for that is the [remote messaging service](https://developer.android.com/reference/android/app/Service#remote-messenger-service-sample).

enguerrand reviewed (commented):
Thanks again for the good work.
Overall this looks mergeable to me (at least to the feature branch).
I am putting this on my phone to try it out for a couple of days and will report back after that.

enguerrand commented on app/src/main/java/de/rochefort/childmonitor/VolumeHistory.java:
```diff
@@ -0,0 +1,65 @@
+        return mHistory.size();
+    }
+
+    protected synchronized void addLast(double volume) {
```
Just found out I can directly commit to your PR, let me know what you think about my approach to handle the concurrency problem here.

fwiesel replied:
Looks great, I didn't know that was possible.

enguerrand commented on app/src/main/java/de/rochefort/childmonitor/VolumeView.java:
```diff
@@ -49,37 +44,25 @@ public VolumeView(Context context, @Nullable AttributeSet attrs, int defStyleAtt
+        final int height = canvas.getHeight();
+        final int width = canvas.getWidth();
+
+        final VolumeHistory history = volumeHistory.getSnapshot(width);
```
Just noticed that your changes to the graph drawing broke the scrolling once the full canvas width is reached.
Not sure what the purpose of these changes are?
Also note that the snapshotting is no longer needed as we now ensure that the volume history is written to on the main event loop, so reading from it within onDraw is thread safe now.

fwiesel replied:
> Not sure what the purpose of these changes are?

It moves more logic out of the loop.

Since `volumeHistory.size() <= width`:
-  `leftMost = 0`.
- and we can iterate until `volumeHistory.size()`

```
            if (i == leftMost (== 0)) {
                xPrev = xNext;
            } 
```
Can be moved out of the loop, if we start with `1`, and initialize the `x/yPrev` values with the ones for `0`, and copy them from the `x/yNext` at the end of the loop.
We loose by that drawing a line for `(i=0)` to `(i=0)`

Finally: `/ maxVolume * (graphHeight) -> * graphHeight / maxVolume -> * grapScale`
 
I think, the bug is actually [here](https://github.com/enguerrand/child-monitor/pull/9/files#diff-34a7e83689bb93f2560f06c0fa9615b983b70c576eaded6d86313f08e89e473cR67-R69). That needs to be done the way you did the `leftMost`logic in the painter.

enguerrand replied:
It also used the wrong history instance further down, but I fixed that too now.

enguerrand replied:
I couldn't resist to remove the snapshot stuff. Sorry for committing one file at a time, I can't convince the Github UI to change two files in one commit ^^

fwiesel commented on app/src/main/java/de/rochefort/childmonitor/ListenService.java:
```diff
@@ -0,0 +1,250 @@
+        String _name = bundle.getString("name");
+
+        // The PendingIntent to launch our activity if the user selects this notification
+        PendingIntent contentIntent = PendingIntent.getActivity(this, 0,
```
Done

fwiesel commented:
I've added another commit. I think, the error message needs to be set in the UI-Thread as well, because
> Only the original thread that created a view hierarchy can touch its views.

enguerrand commented:
> I've added another commit. I think, the error message needs to be set in the UI-Thread as well, because
> 
> > Only the original thread that created a view hierarchy can touch its views.

Yap, makes sense! Thanks!

enguerrand commented:
Today I tested on a different device (running Android 9) and it reproducibly crashed because it attempted to fetch the first element from the VolumeHistory before any elements were added.

I added a commit to fix this.

fwiesel commented:
I have added a bigger change: Splitting the streaming logic out of MonitorActivity into a MonitorService.

I had a device running as a child monitor with power save mode on, and the child monitor stopped immediately when I just switched to another app (on Android 12). This should fix it.

enguerrand commented:
Wow, this is funny, I was talking to wife today about wanting to do this - so this is really appreciated!

However, it needs to be tested and polished a bit more. E.g. When you close it and then reopen through the notification bar entry the UI is not properly repopulated with the device name etc.

Also I managed to prodce the following exception, but did not yet identify the procedure to reproduce it:
```
2024-02-19 21:39:28.719  3441-6857  MediaProvider           com...ndroid.providers.media.module  W  Failed to fetch original file descriptor
java.io.IOException: stat failed: EACCES (Permission denied)
        at com.android.providers.media.MediaProvider.getFileFromFileDescriptor(MediaProvider.java:5794)
        at com.android.providers.media.MediaProvider.getOriginalMediaFormatFileDescriptor(MediaProvider.java:5711)
        at com.android.providers.media.MediaProvider.openTypedAssetFileCommon(MediaProvider.java:7134)
        at com.android.providers.media.MediaProvider.openTypedAssetFile(MediaProvider.java:7115)
        at android.content.ContentProvider$Transport.openTypedAssetFile(ContentProvider.java:548)
        at android.content.ContentProviderNative.onTransact(ContentProviderNative.java:327)
        at android.os.Binder.execTransactInternal(Binder.java:1184)
        at android.os.Binder.execTransact(Binder.java:1143)
Caused by: android.system.ErrnoException: stat failed: EACCES (Permission denied)
        at libcore.io.Linux.stat(Native Method)
        at libcore.io.ForwardingOs.stat(ForwardingOs.java:853)
        at libcore.io.BlockGuardOs.stat(BlockGuardOs.java:420)
        at libcore.io.ForwardingOs.stat(ForwardingOs.java:853)
        at android.app.ActivityThread$AndroidOs.stat(ActivityThread.java:7742)
        at android.system.Os.stat(Os.java:919)
        at com.android.providers.media.MediaProvider.getFileFromFileDescriptor(MediaProvider.java:5788)
        at com.android.providers.media.MediaProvider.getOriginalMediaFormatFileDescriptor(MediaProvider.java:5711) 
        at com.android.providers.media.MediaProvider.openTypedAssetFileCommon(MediaProvider.java:7134) 
        at com.android.providers.media.MediaProvider.openTypedAssetFile(MediaProvider.java:7115) 
        at android.content.ContentProvider$Transport.openTypedAssetFile(ContentProvider.java:548) 
        at android.content.ContentProviderNative.onTransact(ContentProviderNative.java:327) 
        at android.os.Binder.execTransactInternal(Binder.java:1184) 
        at android.os.Binder.execTransact(Binder.java:1143) 
```
Unfortunately I am not sure I will get the chance to look at this with the thoroughness it deserves in the next days. 

So what I am currently thinking is this: Based on the issue report I am assuming there are at least two more people waiting for the ListenService. I think with a couple of days of testing more I would be comfortable to go ahead and merge this (without the MonitorService split-out), possibly do some minor "matter-of-taste" changes and then push it out to f-droid.

Should we move the MonitorService to a different PR?

fwiesel commented:
Sure, I'll open a separate PR for it.

enguerrand commented:
I did not find any other problems, so I'll go ahead and merge this. Thanks again for this valuable contribution!
